plugins/extraction/haskell.ml

```diff
@@ -64,13 +64,18 @@ let preamble mod_name comment used_modules usf =
   ++
   (if not usf.magic then mt ()
    else
-     str "#ifdef __GLASGOW_HASKELL__" ++ fnl () ++
+     str "#ifdef __GLASGOW_HASKELL__ < 900" ++ fnl () ++
```

Comment:
Does this actually work?  It doesn't seem to be valid in C/C++ according to https://en.cppreference.com/w/cpp/preprocessor/conditional, but I can't find the Haskell documentation

Reply:
In the `Example` on that page, I found the following line. So I guess it is fine.
```
#if !defined(DCBA) && (ABCD < 2*4-3)
```

Reply:
For the syntax, I trusted https://guide.aelve.com/haskell/cpp-vww0qd72.

Reply:
If we go with this PR, please test this (or I'll try to test it in the not too distant future) before merging

Reply:
I tested with ghc 8.4.1 (using `ghc --cpp`) and couldn't get the `#` of `unsafeCoerce#` being parsed (was I missing an option).

In any case, based on @pi8027's comment [here](https://github.com/coq/coq/issues/14256#issuecomment-847703462), I tried to use instead `Unsafe.Coerce.unsafeCoerce` which seems to compile well with ghc 8.4.1, and presumably for ghc >= 9, without having to worry about the exact version of ghc. So, I updated the PR.

(Incidentally, the `#ifdef` should have been a `#if` above, but that's not necessary anymore.)

Reply:
> I tested with ghc 8.4.1 (using `ghc --cpp`) and couldn't get the `#` of `unsafeCoerce#` being parsed (was I missing an option).

`{-# OPTIONS_GHC -cpp -XMagicHash #-}` is needed for that.